lib/ecto/changeset.ex

```diff
@@ -220,38 +220,152 @@ defmodule Ecto.Changeset do
+    type = embed!(types, key, fun)
     current = Map.get(model, key)
+
+    do_process_param(key, param_key, kind, params, type, current, acc)
+  end
+
+  defp process_param(key, kind, params, types, model, acc) do
+    {key, param_key} = cast_key(key)
     type = type!(types, key)
+    current = Map.get(model, key)
 
+    do_process_param(key, param_key, kind, params, type, current, acc)
+  end
+
+  defp do_process_param(key, param_key, kind, params, {:embed, embed},
+                        current, {changes, errors}) do
+    {key,
+     case cast_embed(param_key, embed, params, current) do
+       {:ok, embed_changes} ->
+         {Map.put(changes, key, embed_changes), errors}
+       :missing ->
+         {changes, error_on_nil(kind, key, current, errors)}
+       :invalid ->
+         {changes, [{key, "is invalid"}|errors]}
+     end}
+  end
+
+  defp do_process_param(key, param_key, kind, params, type, current, {changes, errors}) do
     {key,
-      case cast_field(param_key, type, params) do
-        {:ok, ^current} ->
-          {changes, error_on_nil(kind, key, current, errors)}
-        {:ok, value} ->
-          {Map.put(changes, key, value), error_on_nil(kind, key, value, errors)}
-        :missing ->
-          {changes, error_on_nil(kind, key, current, errors)}
-        :invalid ->
-          {changes, [{key, "is invalid"}|errors]}
-      end}
-  end
-
-  defp type!(types, key),
-    do: Map.get(types, key) ||
-          raise ArgumentError, "unknown field `#{key}` (note only fields are supported in cast, associations are not)"
+     case cast_field(param_key, type, params) do
+       {:ok, ^current} ->
+         {changes, error_on_nil(kind, key, current, errors)}
+       {:ok, value} ->
+         {Map.put(changes, key, value), error_on_nil(kind, key, value, errors)}
+       :missing ->
+         {changes, error_on_nil(kind, key, current, errors)}
+       :invalid ->
+         {changes, [{key, "is invalid"}|errors]}
+     end}
+  end
+
+  defp embed!(types, key, fun) do
+    case Map.fetch(types, key) do
+      {:ok, {:embed, embed}} ->
+        {:embed, %{embed | changeset: fun}}
+      {:ok, _} ->
+        raise ArgumentError, "only embedded fields can be given a cast function"
+      :error ->
+        raise ArgumentError, "unknown field `#{key}` (note only fields and " <>
+          "embeds are supported in cast, associations are not)"
+    end
+  end
+
+  defp type!(types, key) do
+    case Map.fetch(types, key) do
+      {:ok, {:embed, _} = embed} ->
+        embed
+      {:ok, type} ->
+        type
+      :error ->
+        raise ArgumentError, "unknown field `#{key}` (note only fields and " <>
+          "embeds are supported in cast, associations are not)"
+    end
+  end
 
   defp cast_key(key) when is_binary(key),
     do: {String.to_atom(key), key}
   defp cast_key(key) when is_atom(key),
     do: {key, Atom.to_string(key)}
 
+  defp cast_embed(param_key, %{cardinality: :one, embed: embed, changeset: fun},
```

Comment:
Let's move this (or part of it) to Ecto.Embedded.cast so we don't end-up polluting changeset too much with embeds concerns and diffs and what not.